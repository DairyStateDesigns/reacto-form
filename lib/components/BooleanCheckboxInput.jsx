import React, { Component } from 'react';
import PropTypes from 'prop-types';
import uniqueId from 'lodash.uniqueid';

import customPropTypes from '../shared/propTypes';

class BooleanCheckboxInput extends Component {
  static isComposableFormInput = true;

  static propTypes = {
    ...customPropTypes.inputs,
    className: PropTypes.string,
    label: PropTypes.node.isRequired,
    value: PropTypes.bool,
    style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
  };

  static defaultProps = {
    className: null,
    isReadOnly: false,
    maxLength: null,
    name: null,
    onChanged() {},
    onChanging() {},
    style: null,
    value: undefined,
  };

  constructor(props) {
    super(props);

    this.state = {
      id: uniqueId('BooleanCheckboxInput_'),
      value: props.value || false,
    };
  }

  componentWillMount() {
    this.handleChange(this.props.value || false);
  }

  componentWillReceiveProps(nextProps) {
    const { value } = this.props;
    const { value: nextValue } = nextProps;

    // Whenever a changed value prop comes in, we reset state to that, thus becoming clean.
    if (value !== nextValue) {
      this.setState({ value: nextValue || false });
      this.handleChange(nextValue);
    }
  }

  onChange = (event) => {
    const { checked } = event.target;
    this.setState({ value: checked });
    this.handleChange(checked);
  };

  getValue() {
    return this.state.value;
  }

  resetValue() {
    this.setState({ value: this.props.value || false });
  }

  handleChange(checked) {
    if (this.lastValue === checked) return;
    this.lastValue = checked;
    const { onChanging, onChanged } = this.props;
    onChanging(checked);
    onChanged(checked);
  }

  // Input is dirty if value prop doesn't match value state. Whenever a changed
  // value prop comes in, we reset state to that, thus becoming clean.
  isDirty() {
    return this.state.value !== this.props.value;
  }

  render() {
    const { className, isReadOnly, label, style } = this.props;
    const { id, value } = this.state;

    return (
      <div className={className} style={style}>
        <label htmlFor={id}><input
          checked={value === true}
          id={id}
          onChange={this.onChange}
          readOnly={isReadOnly}
          type="checkbox"
          value="true"
        /> {label}</label>
      </div>
    );
  }
}

export default BooleanCheckboxInput;
