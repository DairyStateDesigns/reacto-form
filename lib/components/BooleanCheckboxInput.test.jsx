import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import BooleanCheckboxInput from './BooleanCheckboxInput';
import registerInputTests from '../../tests/registerInputTests';

// Run generic Input tests that should pass for any Composable Forms Spec input
registerInputTests({
  component: BooleanCheckboxInput,
  defaultValue: false,
  exampleValueOne: true,
  exampleValueTwo: false,
  mount,
  props: { label: 'LABEL' },
  simulateChanged(wrapper, value) {
    wrapper.find('input').simulate('change', { target: { checked: !!value } });
  },
});

test('renders', () => {
  const component = renderer.create(
    <BooleanCheckboxInput name="test" label="LABEL" />,
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
