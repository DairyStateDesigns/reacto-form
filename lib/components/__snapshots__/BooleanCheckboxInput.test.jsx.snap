// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className={undefined}
  style={Object {}}
>
  <label
    htmlFor="BooleanCheckboxInput_8"
  >
    <input
      checked={false}
      id="BooleanCheckboxInput_8"
      onChange={[Function]}
      readOnly={false}
      type="checkbox"
      value="true"
    />
     
    LABEL
  </label>
</div>
`;
